ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,394 @@
+
+In the domain of stream-processing/event-sourcing, this kind of rebuilding the state is known as **replay**. Comparing to reprocessing, the records (commands or events) are not processed again, only the state changes of the events are applied.
+
+In order to extract the processing from the state changes, the workflow instance records are split into commands and events (currently, we have almost only events). 
+
+Commands | Events 
+---|---
+Element_Activating | Element_Activated 
+Element_Completing | Element_Completed
+Element_Terminating | Element_Terminated
+Event_Occur (?) | Event_Occurred (?)
+/ | Sequence_Flow_Taken
+
+![event-sourcing](ZEP-wf-stream-processing/event-sourcing.png)
+
+The **commands** are processed only on processing and drive the workflow execution. The processing of a command includes:
+* read the state of the workflow instance and related resources (RocksDB)
+* check if the record can be applied (i.e. concurrent update guard)
+* apply the business logic  
+* write updates to the state
+* write events and (optionally) follow-up commands to continue the processing
+* apply side-effects (e.g. send a response)
+
+![command-processing](ZEP-wf-stream-processing/event-sourcing-processing.png)
+
+The **events** are processed only on replay and rebuild the state. The replay of an event includes:
+* write updates to the state
+
+![event-replay](ZEP-wf-stream-processing/event-sourcing-replay.png)
+
+In order to rebuild the state with the events, it is necessary that all state changes from the command processing can be applied with the events only. So, the command processing may not update the state without writing an event that reflects the state changes.   
+
+When the replay of the events is done then the processing of the commands starts again. 
+
+## Examples
+
+The following examples show the sequence of records for a certain situation and explain the differences to the current behavior. 
+
+### Create Workflow Instance
+
+Command to Process | Written Events and Follow-up Commands
+---|---
+WorkflowInstanceCreationIntent:Create | WorkflowInstanceCreationIntent:Created, Process:Element_Activating
+Process:Element_Activating | Process:Element_Activated, StartEvent:Element_Activating
+StartEvent:Element_Activating | StartEvent:Element_Activated, StartEvent:Element_Completing 
+StartEvent:Element_Completing | StartEvent:Element_Completed, Sequence_Flow_Taken, ServiceTask:Element_Activating 
+
+* the command `WorkflowInstanceCreationIntent:Create` is the initial command and start of the workflow instance
+* the event `Sequence_Flow_Taken` is produced by the command `StartEvent:Element_Completing` and not by an explicit command for the sequence flow (like `Take_Sequence_Flow`) because it can't be interrupted (i.e. no wait state)
```

Comment:
This is not necessarily needed. We could also consider producing a `Take_Sequence_Flow` follow-up command that later is processed and leads to both a `Sequence_Flow_Taken` event and a `ServiceTask:ElementActivating` follow-up command

Reply:
Chose was made for Sequence_Flow_Taken. The synchronous nature fits our processing better